File: tests/test_aegis_proof_cells.py
# SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

import hashlib
import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
HTML = ROOT / "console" / "3d" / "aegis-proof-cells.html"
ASSET_DIR = ROOT / "console" / "3d" / "aegis-proof-cells"
REGISTRY = ASSET_DIR / "registry.json"
SCRIPT = ASSET_DIR / "app.mjs"
STYLES = ASSET_DIR / "styles.css"
DOC = ROOT / "docs" / "third-party" / "bricklayer-ai-intake-v1.md"


def _sha(path: Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def test_registry_is_original_defensive_and_fail_closed() -> None:
    registry = json.loads(REGISTRY.read_text(encoding="utf-8"))
    boundary = registry["bricklayer_boundary"]
    authority = registry["authority"]

    assert registry["schema"] == "szl.aegis-proof-cells.registry/v1"
    assert registry["operating_mode"] == "DEFENSIVE_READ_ONLY_INVESTIGATION_PLANNER"
    assert boundary["classification"] == "REFERENCE_ONLY_CLEAN_ROOM"
    assert boundary["official_public_code_repository_found"] is False
    assert boundary["proprietary_implementation_available"] is False
    assert boundary["source_code_copied"] is False
    assert boundary["website_copy_copied"] is False
    assert boundary["visual_identity_copied"] is False
    assert boundary["brand_identity_reused"] is False
    assert boundary["affiliation"] == "NONE"

    assert authority["default_effect"] == "DENY"
    assert authority["external_writes"] == "DISABLED"
    assert authority["effectors"] == []
    assert authority["automatic_retries"] == 0
    assert authority["credentials_accepted"] is False
    assert authority["secrets_persisted"] is False
    assert authority["cross_tenant_access"] == "DENIED"
    assert authority["offensive_intrusion"] == "DENIED"
    assert authority["destructive_remediation"] == "DENIED"
    assert authority["production_authorization"] is False


def test_cells_capsules_and_context_contract_are_complete() -> None:
    registry = json.loads(REGISTRY.read_text(encoding="utf-8"))
    cells = registry["proof_cells"]
    capsules = registry["procedure_capsules"]
    cell_ids = {cell["id"] for cell in cells}

    assert len(cells) == 11
    assert len(cell_ids) == 11
    assert len(capsules) == 6
    assert registry["context_types"] == [
        "EVIDENTIARY",
        "PROCEDURAL",
        "INVESTIGATIVE",
        "DECISION",
        "OUTCOME",
    ]
    assert {capsule["mission"] for capsule in capsules} == {
        "alert-triage",
        "phishing",
        "endpoint",
        "vulnerability",
        "cloud",
        "threat-intel",
    }
    for capsule in capsules:
        assert capsule["cells"]
        assert set(capsule["cells"]) <= cell_ids


def test_public_references_are_bounded_and_no_code_is_claimed() -> None:
    registry = json.loads(REGISTRY.read_text(encoding="utf-8"))
    observations = registry["public_architecture_observations"]
    standards = registry["standards_and_public_code"]

    assert len(observations) >= 6
    assert all(item["reuse_policy"] == "REFERENCE_ONLY" for item in observations)
    assert all(item["source"].startswith("https://www.bricklayer.ai/") for item in observations)
    assert all(item["source_code_copied"] is False for item in standards)
    assert {item["name"] for item in standards} == {
        "Open Cybersecurity Schema Framework",
        "MITRE ATT&CK STIX Data",
        "OpenTelemetry Specification",
        "Open Policy Agent",
    }

    doc = DOC.read_text(encoding="utf-8")
    assert "REFERENCE_ONLY_CLEAN_ROOM" in doc
    assert "No official public Bricklayer implementation repository was identified" in doc
    assert "No source from these projects is copied" in doc


def test_page_is_local_digest_bound_and_accessible() -> None:
    html = HTML.read_text(encoding="utf-8")
    css = STYLES.read_text(encoding="utf-8")
    js = SCRIPT.read_text(encoding="utf-8")

    assert 'data-szl-public-experience-v3="true"' in html
    assert 'data-aegis-proof-cells="v1"' in html
    assert 'class="skip"' in html
    assert 'aria-live="polite"' in html
    assert f"styles.css?v={_sha(STYLES)}" in html
    assert f"app.mjs?v={_sha(SCRIPT)}" in html
    assert _sha(REGISTRY) in html

    assert "https://" not in js
    assert "http://" not in js
    assert "innerHTML" not in js
    assert "eval(" not in js
    assert "new Function" not in js
    assert "credentials: \"same-origin\"" in js
    assert "redirect: \"error\"" in js

    for token in (
        "min-height: 48px",
        "overflow-x: hidden",
        "prefers-reduced-motion",
        "prefers-contrast: more",
        "forced-colors: active",
    ):
        assert token in css


def test_runtime_policy_markers_are_explicit() -> None:
    js = SCRIPT.read_text(encoding="utf-8")
    for token in (
        "TENANT_PASSPORT_REQUIRED",
        "CROSS_TENANT_SCOPE",
        "UNSUPPORTED_DEFENSIVE_MISSION",
        "PROHIBITED_ACTION",
        "EVIDENCE_NOT_FRESH",
        "HUMAN_APPROVAL_REQUIRED",
        "DEFENSIVE_PLAN_READY",
        "external_writes: \"DISABLED\"",
        "effectors: []",
        "production_authorization: false",
        "trust_ceiling: 0.97",
    ):
        assert token in js

    prohibited = re.search(r"const PROHIBITED_TOKENS = \[(.*?)\];", js, re.S)
    assert prohibited is not None
    for required in ("exploit", "exfiltrate", "credential theft", "deploy malware", "destructive"):
        assert required in prohibited.group(1)

File: tests/test_aegis_proof_cells.mjs
// SPDX-License-Identifier: Apache-2.0
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import test from "node:test";

import {
  canonicalJSONString,
  evaluateCase,
  evaluateCaseWithReceipt,
} from "../console/3d/aegis-proof-cells/app.mjs";

const registry = JSON.parse(
  await readFile(new URL("../console/3d/aegis-proof-cells/registry.json", import.meta.url), "utf8"),
);

const baseline = {
  tenant_id: "tenant-a",
  passport_tenant_id: "tenant-a",
  alert_id: "alert-1",
  source: "SIEM",
  severity: "HIGH",
  mission: "phishing",
  requested_action: "investigate",
  evidence_count: 4,
  evidence_fresh: true,
  human_approved: false,
};

test("valid defensive investigation produces a bounded sandbox plan", () => {
  const result = evaluateCase(baseline, registry);
  assert.equal(result.decision.state, "SANDBOX_PLAN");
  assert.equal(result.decision.reason, "DEFENSIVE_PLAN_READY");
  assert.equal(result.decision.evidence_class, "MODELED");
  assert.equal(result.decision.production_authorization, false);
  assert.ok(result.decision.score > 0);
  assert.ok(result.decision.score <= 0.97);
  assert.equal(result.procedure_capsule, "phishing-investigation-v1");
  assert.ok(result.plan.length >= 6);
  assert.equal(result.authority.external_writes, "DISABLED");
  assert.deepEqual(result.authority.effectors, []);
});

test("cross-tenant scope is hard denied", () => {
  const result = evaluateCase(
    { ...baseline, passport_tenant_id: "tenant-b" },
    registry,
  );
  assert.equal(result.decision.state, "DENIED");
  assert.equal(result.decision.reason, "CROSS_TENANT_SCOPE");
  assert.equal(result.decision.score, 0);
});

test("offensive and destructive requests are hard denied", () => {
  for (const action of [
    "exploit the target",
    "exfiltrate evidence",
    "dump credential hashes",
    "deploy malware",
    "destructive cleanup",
  ]) {
    const result = evaluateCase({ ...baseline, requested_action: action }, registry);
    assert.equal(result.decision.state, "DENIED");
    assert.equal(result.decision.reason, "PROHIBITED_ACTION");
    assert.equal(result.decision.score, 0);
  }
});

test("stale evidence abstains and remediation requires approval", () => {
  const stale = evaluateCase(
    { ...baseline, evidence_fresh: false },
    registry,
  );
  assert.equal(stale.decision.state, "ABSTAINED");
  assert.equal(stale.decision.reason, "EVIDENCE_NOT_FRESH");

  const gated = evaluateCase(
    { ...baseline, requested_action: "purge-email", human_approved: false },
    registry,
  );
  assert.equal(gated.decision.state, "AWAITING_APPROVAL");
  assert.equal(gated.decision.reason, "HUMAN_APPROVAL_REQUIRED");

  const approved = evaluateCase(
    { ...baseline, requested_action: "purge-email", human_approved: true },
    registry,
  );
  assert.equal(approved.decision.state, "SANDBOX_PLAN");
  assert.equal(approved.decision.production_authorization, false);
});

test("identical input produces an identical unsigned deterministic receipt", async () => {
  const first = await evaluateCaseWithReceipt(baseline, registry);
  const second = await evaluateCaseWithReceipt(baseline, registry);
  assert.equal(first.proof_chain.sha256, second.proof_chain.sha256);
  assert.equal(first.proof_chain.sha256.length, 64);
  assert.equal(first.proof_chain.signature_status, "UNAVAILABLE");
  assert.equal(first.proof_chain.persisted, false);
  assert.equal(canonicalJSONString(first).includes("undefined"), false);
});
